### PR TITLE
Snapshot code name refactors and comments

### PR DESCRIPTION
**Pull request by DiannaHohensee**
"Renaming some classes/variables to be more descriptive/accurate.
 Also added comments here and there. Added to the package-level
 documentation."

-------------------------

I was investigating the code, and these are comments and renames I added to improve my own understanding.

### CODE REVIEW

**Comment by DiannaHohensee** [comment]
Existing test failure: https://github.com/elastic/elasticsearch/issues/109816

**Review by mark-vieira — APPROVED** [review]

**Review by original-brownbear — CHANGES_REQUESTED** [review]
I don't think this is a good idea to be honest. As @idegtiarenko points out the nested class naming should work fine, no need to repeat the prefix IMO. This also makes the code considerably longer.

But most importantly, this code is really tricky to follow and there are a number of far-from-obvious conditions in the state machine where git-blame helps a lot at times :) This kind of renaming makes following the history just so much harder, I wouldn't do it just for that reason, even if it was a win in isolation.

**Comment by DaveCTurner** [comment]
I think these are _super_ weak arguments against doing this. In some places we call these things by their full name, but oftentimes it's just `Entry`, and it is definitely hard on the reader to keep track of which `Entry` we're talking about. And IMO it's easy to `git blame` across this kind of change in practice, certainly not so difficult that we should preserve this code in its current fairly-unreadable state.

**Comment by DiannaHohensee** [comment]
> It is clear that the Entry belongs to SnapshotsInProgress as it is declared inside.

I have trouble reading code that simply writes `entry`, `state` and `snapshot` everywhere. It's difficult to distinguish Entry and State types when both SnapshotDeletionsInProgress and SnapshotsInProgress have the same Entry and State subclass names 😅 Some code blocks go through so many snapshot related types, it becomes unrealistic to use the IDE to keep checking, and referring back to, types: it's so much easier to just rename them -- I renamed some variable names, too, but it's weird without the type name as well. My mind also tends to slide off of generic/non-descriptive names like Entry: I have to constantly hold what it refers to fixed in my mind, and that's work.

Then SnapshotsInProgress is plural, while the interior Entry class is actually singular, not the entry for the whole SnapshotsInProgress -- confused me initially. Thus part of the reason for an explicit SnapshotInProgressState/Entry singular. 

I ran into naming conflicts when I initially changed State -> SnapshotState with this other [miscellaneous SnapshotState class](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/snapshots/SnapshotState.java), which is unrelated to the snapshot mechanics. Renaming it, to distinguish it, wouldn't hurt, but I didn't investigate what it's used for.

> ...this code is really tricky to follow and there are a number of far-from-obvious conditions in the state machine where git-blame helps a lot at times :)

Git blame has helped me to some small degree (tells me what features/behavior we should have via the commit message), but also not a whole lot because the changes in each patch are quick fixes to code hither and yon that has grown disorganized over time, and there are few code comments. Codepaths are reused for different things, or multiple codepaths are used for the same thing, I'm not sure git-blame is that helpful to the casual reader.

--------------------

In conclusion, these changes help me understand the code, and don't hinder other people understanding the code.

File: modules/data-streams/src/test/java/org/elasticsearch/datastreams/action/DeleteDataStreamTransportActionTests.java [hunk 1]
```diff
@@ -131,12 +131,12 @@ public void testDeleteSnapshottingDataStream() {
         );
     }
 
-    private SnapshotsInProgress.Entry createEntry(String dataStreamName, String repo, boolean partial) {
-        return SnapshotsInProgress.Entry.snapshot(
+    private SnapshotsInProgress.SnapshotInProgressEntry createEntry(String dataStreamName, String repo, boolean partial) {
+        return SnapshotsInProgress.SnapshotInProgressEntry.snapshot(
```
[the remark this answers is not in the document]
Comment (DiannaHohensee) [reply modules/data-streams/src/test/java/org/elasticsearch/datastreams/action/DeleteDataStreamTransportActionTests.java:135]: What makes this hard for me is that `Entry` is not only used inside `SnapshotsInProgress.java`, so it leaks out and becomes harder to reason about in wider contexts.
